consensus/src/dag/bootstrap.rs

```diff
@@ -164,40 +165,89 @@ impl TDagMode for SyncMode {
-            },
-            Err(_) => {
-                info!("sync failed. continuing without advancing.");
-                false
-            },
-        };
+        let (request, responders, sync_dag_store) = sync_manager.build_sync_to_request(
+            &self.certified_node_msg,
+            self.base_state.dag_store.clone(),
+            highest_committed_anchor_round,
+        );
 
-        if success {
-            let (new_state, new_handler, new_fetch_service) = bootstrapper.full_bootstrap();
-            Mode::Active(ActiveMode {
-                handler: new_handler,
-                fetch_service: new_fetch_service,
-                base_state: new_state,
-            })
-        } else {
-            let (new_handler, new_fetch_service) =
-                bootstrapper.bootstrap_components(&self.base_state);
-            Mode::Active(ActiveMode {
-                handler: new_handler,
-                fetch_service: new_fetch_service,
-                base_state: self.base_state,
-            })
+        let commit_li = self.certified_node_msg.ledger_info().clone();
+
+        let network_handle = SyncModeMessageHandler::new(
+            bootstrapper.epoch_state.clone(),
+            request.start_round(),
+            request.target_round(),
+        );
+
+        let (shutdown_handle, shutdown) = shutdown_handle.new_child();
+
+        let (res_tx, res_rx) = oneshot::channel();
+        tokio::spawn(async move {
+            let success = match sync_manager
+                .sync_dag_to(
+                    dag_fetcher,
+                    request,
+                    responders,
+                    sync_dag_store,
+                    commit_li,
+                    shutdown,
+                )
+                .await
+            {
+                Ok(_) => {
+                    info!("sync success. going to rebootstrap.");
+                    true
+                },
+                Err(_) => {
+                    info!("sync failed. continuing without advancing.");
+                    false
+                },
+            };
+            let _ = res_tx.send(success);
+        });
+
+        let mut buffer = Vec::new();
+
+        select! {
+            res = network_handle.run(dag_rpc_rx, &mut buffer) => {
+                // The network handle returns if the sender side of dag_rpc_rx closes,
+                // or network handle found a future CertifiedNodeMessage to cancel the
+                // current sync.
+                if let Some(msg) = res {
+                    shutdown_handle.shutdown().await;
+                    return Mode::Sync(SyncMode {
+                        certified_node_msg: msg,
+                        base_state: self.base_state,
+                    });
+                } else {
+                    // Usually happens if the sender side of dag_rpc_rx closes first, which
+                    // shouldn't happen.
+                    return Mode::Exit(ExitMode {  })
```

Comment:
I don't feel we need this mode, `run` can return an Option and we can select on Some(next_mode)?